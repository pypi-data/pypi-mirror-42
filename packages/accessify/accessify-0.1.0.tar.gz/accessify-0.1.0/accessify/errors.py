"""
Provide errors for accessibility levels.
"""
INACCESSIBLE_DUE_TO_ITS_PROTECTION_LEVEL_EXCEPTION_MESSAGE = \
    '{class_name}.{method_name}() is inaccessible due to its protection level'


class InaccessibleDueToItsProtectionLevelException(Exception):
    """
    Inaccessible due to its protection level exception.
    """

    def __init__(self, message):
        self.message = message