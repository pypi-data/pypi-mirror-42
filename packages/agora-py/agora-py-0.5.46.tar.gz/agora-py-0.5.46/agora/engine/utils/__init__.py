"""
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Ontology Engineering Group
        http://www.oeg-upm.net/
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Copyright (C) 2016 Ontology Engineering Group.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

            http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
#-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=#
"""
import os
from threading import Lock

__author__ = 'Fernando Serena'


class Singleton(type):
    _lock = Lock()

    def __call__(cls, *args, **kwargs):
        with cls._lock:
            id = kwargs.get('id', '')
            if id not in cls.instances:
                cls.instances[id] = super(Singleton, cls).__call__(*args, **kwargs)
            return cls.instances[id]


def prepare_store_path(base, path):
    if base and not os.path.exists(base):
        os.makedirs(base)
    if base or path:
        full_path = '/'.join(filter(lambda x: x, [base, path]))
        if not os.path.exists(full_path):
            os.makedirs(full_path)
        return full_path


class Semaphore(object):
    def __init__(self):
        self.value = -1

    def __enter__(self):
        self.value = 0

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.value = 1

    def isSet(self):
        return self.value == 1

    def set(self):
        self.value = 1


def get_immediate_subdirectories(a_dir):
    if not os.path.exists(a_dir):
        os.makedirs(a_dir)
    return [name for name in os.listdir(a_dir)
            if os.path.isdir(os.path.join(a_dir, name))]


class Wrapper(object):
    def __init__(self, object):
        self.__lock = Lock()
        self.__attr_locks = {}

        if isinstance(object, Wrapper):
            self.__object = object.__object
            self.__argret = object.__argret
            self.__attr_locks = object.__attr_locks
            self.__lock = object.__lock
        else:
            self.__object = object
            self.__argret = {}

    def decorate(self, f):
        def wrapper(*args, **kwargs):

            key = f.func_name + str(args)
            with self.__lock:
                if key not in self.__attr_locks:
                    self.__attr_locks[key] = Lock()

            with self.__attr_locks[key]:
                if key not in self.__argret:
                    ret = f(*args, **kwargs)

                    if kwargs:
                        return ret
                    if ret is not None:
                        self.__argret[key] = ret

                return self.__argret.get(key, None)

        return wrapper

    def __getattr__(self, item):

        if item.startswith('__'):
            return self.__object.__getattribute__(item)

        with self.__lock:
            if item not in self.__attr_locks:
                self.__attr_locks[item] = Lock()

        with self.__attr_locks[item]:
            if item not in self.__argret:
                attr = self.__object.__getattribute__(item)
                if hasattr(attr, '__call__'):
                    return self.decorate(attr)
                else:
                    self.__argret[item] = attr

        return self.__argret[item]


stopped = Semaphore()
