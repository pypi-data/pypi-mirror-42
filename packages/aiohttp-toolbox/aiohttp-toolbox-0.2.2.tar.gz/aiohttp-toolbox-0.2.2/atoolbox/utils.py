import json
import re
from typing import Any, Tuple, Type, TypeVar, Union

from aiohttp.web import Response
from aiohttp.web_exceptions import HTTPException
from pydantic import BaseModel, ValidationError, validate_model
from pydantic.fields import Shape

from .json_tools import pretty_lenient_json

JSON_CONTENT_TYPE = 'application/json'
IP_HEADER = 'X-Forwarded-For'
PROTO_HEADER = 'X-Forwarded-Proto'
URI_NOT_ALLOWED = re.compile(r'[^a-zA-Z0-9_\-/.]')
REMOVE_PORT = re.compile(r':\d{2,}$')
PydanticModel = TypeVar('PydanticModel', bound=BaseModel)

__all__ = (
    'raw_json_response',
    'json_response',
    'parse_request_json',
    'parse_request_json_ignore_missing',
    'parse_request_query',
    'get_ip',
    'request_root',
    'JsonErrors',
    'encrypt_json',
    'decrypt_json',
    'get_offset',
    'slugify',
    'remove_port',
    'RequestError',
)


def raw_json_response(json_str: Union[str, bytes, None], status_: int = 200):
    if isinstance(json_str, str):
        body = json_str.encode()
    elif isinstance(json_str, bytes):
        body = json_str
    elif json_str is None:
        body = b'null'
    else:
        raise TypeError(f'json_str must be bytes or str, not "{type(json_str)}')
    return Response(body=body + b'\n', status=status_, content_type=JSON_CONTENT_TYPE)


def json_response(*, status_=200, list_=None, headers_=None, **data):
    return Response(
        body=json.dumps(data if list_ is None else list_).encode() + b'\n',
        status=status_,
        content_type=JSON_CONTENT_TYPE,
        headers=headers_,
    )


async def parse_request_json(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    error_details = None
    try:
        data = await request.json()
    except ValueError:
        error_msg = 'Invalid JSON'
    else:
        try:
            return model.parse_obj(data)
        except ValidationError as e:
            error_msg = 'Invalid Data'
            error_details = e.errors()

    raise JsonErrors.HTTPBadRequest(message=error_msg, details=error_details, headers=headers)


def parse_request_query(request, model: Type[PydanticModel], *, headers=None) -> PydanticModel:
    data = {}
    for k in request.query:
        v = request.query.getall(k)
        f = model.__fields__.get(k)
        if len(v) > 1 or f and f.shape != Shape.SINGLETON:
            data[k] = v
        else:
            data[k] = v[0]

    try:
        return model(**data)
    except ValidationError as e:
        raise JsonErrors.HTTPBadRequest(message='Invalid query data', details=e.errors(), headers=headers)


async def parse_request_json_ignore_missing(
    request, model: Type[PydanticModel], *, headers=None
) -> Tuple[PydanticModel, dict]:
    try:
        raw_data = await request.json()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(message='Invalid JSON', headers=headers)
    if not isinstance(raw_data, dict):
        raise JsonErrors.HTTPBadRequest(message='data not a dictionary', headers=headers)

    data, e = validate_model(model, raw_data, raise_exc=False)
    if e:
        errors = [e for e in e.errors() if not (e['type'] == 'value_error.missing' and len(e['loc']) == 1)]
        if errors:
            raise JsonErrors.HTTPBadRequest(message='Invalid Data', details=errors, headers=headers)

    return model.construct(**data), raw_data


def get_ip(request):
    ips = request.headers.get(IP_HEADER)
    if ips:
        return ips.split(',', 1)[0].strip(' ')
    else:
        return request.remote


def request_root(request):
    # request.url.scheme doesn't work as https is already terminated
    scheme = request.headers.get(PROTO_HEADER) or 'http'
    return f'{scheme}://{request.host}'


class JsonErrors:
    class _HTTPExceptionJson(HTTPException):
        custom_reason = None

        def __init__(self, message, *, details=None, headers=None):
            self.message = message
            self.details = details
            data = {'message': message}
            if details:
                data['details'] = details
            super().__init__(
                text=pretty_lenient_json(data),
                content_type=JSON_CONTENT_TYPE,
                headers=headers,
                reason=self.custom_reason,
            )

        def __repr__(self) -> str:
            return f'{super().__repr__()}, {self.status}: {self.message}'

        def __str__(self) -> str:
            return repr(self)

    class HTTPAccepted(_HTTPExceptionJson):
        status_code = 202

    class HTTPBadRequest(_HTTPExceptionJson):
        status_code = 400

    class HTTPUnauthorized(_HTTPExceptionJson):
        status_code = 401

    class HTTPPaymentRequired(_HTTPExceptionJson):
        status_code = 402

    class HTTPForbidden(_HTTPExceptionJson):
        status_code = 403

    class HTTPNotFound(_HTTPExceptionJson):
        status_code = 404

    class HTTPMethodNotAllowed(_HTTPExceptionJson):
        status_code = 405

        def __init__(self, message, allowed_methods, *, headers=None):
            headers = headers or {}
            headers.setdefault('Allow', ','.join(allowed_methods))
            super().__init__(message, details={'allowed_methods': allowed_methods}, headers=headers)

    class HTTPConflict(_HTTPExceptionJson):
        status_code = 409

    class HTTP470(_HTTPExceptionJson):
        status_code = 470
        custom_reason = 'Invalid user input'


def encrypt_json(app, data: Any) -> str:
    return app['auth_fernet'].encrypt(json.dumps(data).encode()).decode()


def decrypt_json(app, token: bytes, *, ttl: int = None, headers=None) -> Any:
    from cryptography.fernet import InvalidToken

    try:
        return json.loads(app['auth_fernet'].decrypt(token, ttl=ttl).decode())
    except InvalidToken:
        raise JsonErrors.HTTPBadRequest(message='invalid token', headers=headers)


def get_offset(request, paginate_by=100):
    page = request.query.get('page')
    if not page:
        return 0

    try:
        p = int(page)
        if p < 1:
            raise ValueError()
    except ValueError:
        raise JsonErrors.HTTPBadRequest(f"invalid page '{page}'")
    else:
        return (p - 1) * paginate_by


def slugify(title):
    name = title.replace(' ', '-').lower()
    name = URI_NOT_ALLOWED.sub('', name)
    name = re.sub('-{2,}', '-', name)
    return name.strip('_-')


def remove_port(host):
    return REMOVE_PORT.sub('', host)


class RequestError(RuntimeError):
    def __init__(self, status, url, *, text: str = None):
        self.status = status
        self.url = url
        self.text = text

    def __str__(self):
        return f'response {self.status} from "{self.url}"' + (f':\n{self.text[:400]}' if self.text else '')

    def json(self):
        return json.loads(self.text)

    def extra(self):
        return self.text
