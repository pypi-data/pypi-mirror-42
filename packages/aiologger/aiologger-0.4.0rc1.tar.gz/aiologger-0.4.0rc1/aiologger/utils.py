from typing import Callable


class classproperty:
    def __init__(self, func):
        self._func = func

    def __get__(self, obj, owner):
        return self._func(owner)


class CallableWrapper:
    def __init__(self, func: Callable) -> None:
        self.func = func

    def __call__(self, *args, **kwargs):
        return self.func(*args, **kwargs)
