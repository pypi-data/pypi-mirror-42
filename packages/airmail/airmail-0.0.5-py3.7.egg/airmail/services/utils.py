import os, sys
from pathlib import Path
import click
import yaml

def err_log(msg, *args):
    if args:
        msg %= args
    click.echo(click.style('Error: ', fg='red', bold=True) + msg)

# Read a yaml file given a file path. Shou
def read_yml(filepath):
    target_file = Path(filepath)

    if target_file.is_file():
        with open(target_file) as yml_file:
            return yaml.load(yml_file)
    else:
        click.echo(click.style('Error: ', fg='red', bold=True) +
            'Could not read the .yml at ' +
            click.style(filepath, bold=True))
        sys.exit()

# Determines the path to files inside the
# project. (i.e. the `data` directory)
def determine_project_path():
    """Borrowed from wxglade.py"""
    try:
        root = __file__
        if os.path.islink (root):
            root = os.path.realpath (root)
        return os.path.dirname (os.path.abspath (root))
    except:
        print("I'm sorry, but something is wrong.")
        print("There is no __file__ variable. Please contact the author.")
        sys.exit ()
