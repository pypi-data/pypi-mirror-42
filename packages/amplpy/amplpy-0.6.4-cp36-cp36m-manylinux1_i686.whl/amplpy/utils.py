# -*- coding: utf-8 -*-
from __future__ import print_function, absolute_import, division
from builtins import map, range, object, zip, sorted
from past.builtins import basestring, unicode
from sys import version_info

from . import amplpython
from .base import BaseClass


def lock_and_call(call, lock):
    lock.acquire()
    try:
        result = call()
    except Exception:
        lock.release()
        raise
    else:
        lock.release()
        return result


class Tuple(BaseClass):
    def __init__(self, values=None, **kwargs):
        if values is not None:
            values = Utils.convToList(values)
            va = Utils.toVariantArray(values)
            self._impl = amplpython.Tuple.Factory(va, len(values))
        else:
            self._impl = kwargs.get('_impl', None)

    def toTuple(self):
        return tuple(
            Utils.castVariant(self._impl.getIndex(i))
            for i in range(self._impl.size())
        )

    def toPyObj(self):
        size = self._impl.size()
        if size == 1:
            return Utils.castVariant(self._impl.getIndex(0))
        else:
            return self.toTuple()

    @classmethod
    def fromTupleRef(cls, tupleRef):
        return cls(_impl=amplpython.Tuple.Factory(tupleRef))


class Utils(object):
    @staticmethod
    def toVariantArray(lst):
        va = amplpython.VariantArray(len(lst))
        for i in range(len(lst)):
            if isinstance(lst[i], unicode):
                va[i] = amplpython.Variant(str(lst[i]))
                # FIXME: This is just a workaround for issue amplapi#332
                # The real fix requires a new release of amplapi
            else:
                va[i] = amplpython.Variant(lst[i])
        return va

    @staticmethod
    def toTupleArray(lst):
        ta = amplpython.TupleArray(len(lst))
        for i in range(len(lst)):
            ta[i] = Tuple(lst[i])._impl
        return ta

    @staticmethod
    def castVariant(variant):
        if variant.type() == amplpython.NUMERIC:
            return variant.dbl()
        elif variant.type() == amplpython.STRING:
            return variant.str()

    @staticmethod
    def castStringArray(sarray):
        return [sarray.getIndex(i) for i in range(sarray.size())]

    @staticmethod
    def castVariantRef(variantref):
        return Utils.castVariant(amplpython.Variant(variantref))

    @staticmethod
    def convToList(value):
        if isinstance(value, list):
            return value
        elif isinstance(value, tuple):
            return list(value)
        else:
            return [value]


def multidict(d):
    try:
        ncols = min(len(d[k]) for k in d)
    except TypeError:
        raise TypeError
    return [list(d.keys())] + [{k: d[k][i] for k in d} for i in range(ncols)]
