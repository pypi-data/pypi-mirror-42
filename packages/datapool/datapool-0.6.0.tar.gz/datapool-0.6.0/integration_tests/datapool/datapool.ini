[db]
connection_string = sqlite+pysqlite:////Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz/.simple.db

[backup_landing_zone]
folder = 

[r]
extension = .r
executable = /usr/local/bin/R

[matlab]
extension = .m
executable = /Applications/MATLAB_R2016a.app/bin/matlab

[julia]
extension = .jl
version = 0.5.2
executable = /usr/local/bin/julia

[python]
extension = .py
executable = /Users/uweschmitt/Projects/eawag_datapool/venv/bin/python3

[logging]
config_file = ./logging_config.yaml

[log_receiver]
port = 5559

[worker]
port = 5555
count = 5

[server]
pid_file = /Users/uweschmitt/.datapool/datapool.pid

[landing_zone]
folder = /Users/uweschmitt/Projects/eawag_datapool/src/integration_tests/lz

