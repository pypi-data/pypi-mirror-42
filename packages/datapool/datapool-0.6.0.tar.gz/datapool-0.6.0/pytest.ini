[pytest]
addopts = -rs --full-trace
testpaths = tests
