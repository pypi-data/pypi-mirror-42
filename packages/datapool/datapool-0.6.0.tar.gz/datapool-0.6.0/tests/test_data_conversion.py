# encoding: utf-8
from __future__ import absolute_import, division, print_function

import os
from collections import defaultdict

import pytest

from datapool.data_conversion import ConversionRunner
from datapool.errors import FormatError, InvalidOperationError
from datapool.utils import find_executable

has_matlab = find_executable("matlab") is not None


def _run_conversion(interpreter_name, script, data_path, config):
    folder = data_path("test_landing_zone/data/FloDar/FloDar_%s/" % interpreter_name)
    raw_data_path = os.path.join(folder, "raw_data", "data-001.raw")
    # section names are lower case:
    getattr(config, interpreter_name.lower()).executable = find_executable(
        interpreter_name
    )
    return ConversionRunner(config).run_conversion(script, raw_data_path)


def _run_matlab_conversion(script, data_path, config):
    return _run_conversion("matlab", script, data_path, config)


def _run_julia_conversion(script, data_path, config):
    return _run_conversion("julia", script, data_path, config)


def _run_python_conversion(script, data_path, config):
    return _run_conversion("python", script, data_path, config)


def _run_r_conversion(script, data_path, config):
    return _run_conversion("R", script, data_path, config)


def _dump_rows(rows, regtest_print):

    regtest_print("converted data file has {} rows".format(len(rows)))

    groups = defaultdict(list)
    for row in rows:
        groups[row["parameter"]].append(row)

    def p(row_dict):
        line = (
            "{"
            + ", ".join("{}: {!r}".format(k, v) for (k, v) in sorted(row_dict.items()))
            + "}"
        )
        regtest_print(line)

    for parameter in sorted(groups.keys()):
        rows = groups[parameter]
        for row in rows[:5]:
            p(row)
        regtest_print("...")
        for row in rows[-5:]:
            p(row)
        regtest_print()


@pytest.mark.skipif(not has_matlab, reason="no matlab interpreter found on your system")
def test_matlab_file_conversion(data_path, setup_logger, regtest_print, config):

    script = data_path("test_landing_zone/data/FloDar/FloDar_matlab/conversion.m")
    rows = _run_matlab_conversion(script, data_path, config)
    _dump_rows(rows, regtest_print)


@pytest.mark.skipif(not has_matlab, reason="no matlab interpreter found on your system")
def test_matlab_file_conversion_failing(data_path, config, regtest_logger):

    script = os.path.join(data_path("failing_scripts", "conversion.m"))

    with regtest_logger():
        with pytest.raises(InvalidOperationError):
            _run_matlab_conversion(script, data_path, config)


def test_julia_file_conversion(data_path, setup_logger, config, regtest_print):

    script = data_path("test_landing_zone/data/FloDar/FloDar_julia/conversion.jl")

    rows = _run_julia_conversion(script, data_path, config)
    _dump_rows(rows, regtest_print)


def test_julia_file_conversion_failing(
    data_path, config, regtest_logger, regtest_print
):

    script = data_path("failing_scripts", "conversion_exceeds_dimensions.jl")

    with regtest_logger():
        with pytest.raises(InvalidOperationError) as e:
            _run_julia_conversion(script, data_path, config)
        regtest_print(e.value)


def test_julia_file_wrong_function_name(
    data_path, config, regtest_logger, regtest_print
):

    script = data_path("failing_scripts", "wrong_function_name.jl")

    with regtest_logger():
        with pytest.raises(InvalidOperationError) as e:
            _run_julia_conversion(script, data_path, config)
        regtest_print(e.value)


def test_julia_file_wrong_module_name(data_path, config, regtest_logger, regtest_print):

    script = data_path("failing_scripts", "wrong_module_name.jl")
    with regtest_logger():
        with pytest.raises(InvalidOperationError) as e:
            _run_julia_conversion(script, data_path, config)
        regtest_print(e.value)


def test_julia_file_non_ascii_output(data_path, config, regtest_logger, regtest_print):

    script = data_path("failing_scripts", "non_ascii_output.jl")

    config.julia.executable = find_executable("julia")

    with regtest_logger():
        with pytest.raises(FormatError) as e:
            _run_julia_conversion(script, data_path, config)
        regtest_print(e.value)


def test_python_file_conversion(data_path, setup_logger, config, regtest_print):

    script = data_path("test_landing_zone/data/FloDar/FloDar_python/conversion.py")

    rows = _run_python_conversion(script, data_path, config)
    _dump_rows(rows, regtest_print)


def test_r_file_conversion(data_path, setup_logger, config, regtest_print):

    script = data_path("test_landing_zone/data/FloDar/FloDar_R/conversion.r")

    rows = _run_r_conversion(script, data_path, config)
    _dump_rows(rows, regtest_print)
