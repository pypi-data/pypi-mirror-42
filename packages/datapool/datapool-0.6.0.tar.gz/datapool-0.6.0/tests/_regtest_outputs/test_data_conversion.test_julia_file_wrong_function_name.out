ERROR    eawag.datapool  JULIA: include("tests/data/failing_scripts/wrong_function_name.jl")
ERROR    eawag.datapool  JULIA: wrong_function_name.convert("<tmpdir_from_tempfile_module>/tests__data__failing_scripts__wrong_function_name_jl/raw_data_1.txt", "<tmpdir_from_tempfile_module>/tests__data__failing_scripts__wrong_function_name_jl/out_data_1.txt")
ERROR    eawag.datapool  JULIA: 
ERROR    eawag.datapool  JULIA: MethodError: First argument to `convert` must be a Type, got <tmpdir_from_tempfile_module>/tests__data__failing_scripts__wrong_function_name_jl/raw_data_1.txt
ERROR    eawag.datapool  JULIA: 
running script failed. exit_code is 1.
